atomics/include/desul/atomics/SYCL.hpp

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright (c) 2019, Lawrence Livermore National Security, LLC
+and DESUL project contributors. See the COPYRIGHT file for details.
+Source: https://github.com/desul/desul
+
+SPDX-License-Identifier: (BSD-3-Clause)
+*/
+#ifndef DESUL_ATOMICS_SYCL_HPP_
+#define DESUL_ATOMICS_SYCL_HPP_
+
+#ifdef DESUL_HAVE_SYCL_ATOMICS
+#include "desul/atomics/Common.hpp"
+
+namespace desul {
+namespace Impl {
+template<class T>
+struct is_sycl_atomic_type {
+  static constexpr bool value = std::is_same<T, int>::value ||
+                                std::is_same<T, unsigned int>::value ||
+				std::is_same<T, long>::value ||
+				std::is_same<T, unsigned long>::value ||
+				std::is_same<T, long long>::value ||
+                                std::is_same<T, unsigned long long int>::value ||
+				std::is_same<T, float>::value ||
+				std::is_same<T, double>::value;
+};
+} // Impl
+
+// Atomic Add
+template<class T, class MemoryOrder/*, class MemoryScope*/>
+inline
+typename std::enable_if<Impl::is_sycl_atomic_type<T>::value,T>::type
+atomic_fetch_add(T* dest, T val, MemoryOrder, MemoryScopeDevice) {
+  DESUL_SYCL_NAMESPACE::atomic_ref<
+    T,
+    DesulToSYCLMemoryOrder<MemoryOrder>::value,
+    DesulToSYCLMemoryScope<MemoryScopeDevice>::value,
+    sycl::access::address_space::global_device_space>
```

Comment:
Consider doing
```C++
namespace Impl {
template <class T, class MemoryOrder, class MemoryScope>
using sycl_atomic_ref = DESUL_SYCL_NAMESPACE:: atomic_ref<T, DesulToSYCLMemoryOrder<MemoryOrder>::value, DesulToSYCLMemoryScope<MemoryScopeDevice>::value, AppropriateAddressSpace>;

Reply:
Can you please comment on the address space.  I see it in the DPC++ doc but I could not find the `global_device_space` enumerator  in the SYCL specification.

Reply:
Since the current implementation only works for device scope, this seemed to be the appropriate `access_space`. The values allowed for `atomic_ref` in the release I am using right now are `global_space, local_space, global_device_space`.

The `SYCL2020` specifications say that the default value should be `generic_space` and defined values are `global_space, local_space, constant_space, private_space, generic_space`.

The values defined in the DPC++ implementation are `private_space, global_space, constant_space, local_space, global_device_space, global_host_space`.

Reply:
This concerns me.  Are you saying that the atomics do not work on local memory or private memory? What happens if you specify generic memory?  I would want a pretty substantial warning to users if we can only support device memory atomics and not local or private atomics since we can't be sure where the pointers the functions are being used on are located.

Reply:
The specifications say
>  For  performance  tuning  and  genericness.  Even  if  the  platform  supports  the  representation  of  thegeneric address space, this may come at some performance sacrifice. In order to help the target com­piler, it can be useful to track specifically which address space a pointer is addressing.

At the moment, the `DPC++` implementation doesn't support `generic_space`, though, (see above).

Reply:
That's extremely unfortunate, especially since the interface we have can't express the difference between local and global memory pointers, let alone shared memory pointers from the host.  If the only option we have is to use the global space, then so be it.  Will have to think about whether we can, or want to, do anything to be able to express this, but for now I'd say make it generic unless you're on dpc+, then whatever is least broken there.

Reply:
`dpcpp` is the only compiler I have available that supports SYCL so I am hesitant to specify something that can't be tested (despite being in the specifications).